### ISSUE TITLE
Add tests to verify the visuals and functionality of generated VB code matches that of C#

### ISSUE THREAD

**Issue by mrlacey** [Microsoft/WindowsTemplateStudio#1336]
Follow on from #863 

Depends on #916 

Check equivalent generated project create the same visual output (via WAD)

**Comment by mrlacey** [comment]
Implementing this with a temporary workaround that doesn't require #916 doing first.

**Comment by mrlacey** [comment]
Changes for #1479 are going to have a big impact here...

**Comment by ralarcon** [comment]
@mrlacey could you help me better understand the impact?

**Comment by mrlacey** [comment]
@ralarcon I'm currently looking to generate a base UI test project and then include different files for different types of tests. At a fundamental level, I'm doing something that wasn't intended to be as automated as I need and that means the whole thing's a bit more complicated than I was hoping. On top of that I'm loading new templates in a way that may now need to change based on what happens with #1479. However, if I can just tell the templating engine to "just look over here as well" then that might be simple. More investigation needed once #1479 is further along. 

My comment above was just a note to myself and an appropriate link to cross reference the issues.

**Comment by ralarcon** [comment]
That's fair. Lets progress on #1479 and then will be easy!

**Comment by mrlacey** [comment]
I currently have a few pages fully working. Only these are working as their content is fixed. Moving onto other pages next.

Here's what the tests do
- Check WinAppDriver is installed
- For each project type and each framework supported by both languages
- For each page
- Generate the C# app with just that page
- Make sure that is the page that is shown when the app launches
- Disable DotNetNative compilation (so release builds can be created faster)
- Do a release build
- Install the Appx generated by the release build
- Get the PFN of the installed app (not available from the source code)
- Install the temporary certificate the Appx was generated with (without this can't run the app)
- Repeat the above (generate, install, etc.) for a VB version of the same app
- Create project with appropriate tests
- Update test project with PFNs of installed apps
- Build test project
- Run all tests in test project from the command line
- Tests do the following
-- Launch app
-- Maximize screen (so all images are the same size and position)
-- Take & save screenshot
-- Close app
-- Repeat for app 2
-- Compare the generated images
--- Get a smaller version of each image (so comparison is faster)
--- Convert image to grayscale (so only need compare the Red byte of each pixel)
--- Compare every byte in image (but exclude area containing the app title)
-- If images do not match then a diff image is created
- If tests pass then everything is uninstalled & deleted 
- If the test fails then everything is left behind for investigation


When a test fails, a "diff image" is generated that highlights the area where differences between the two images by drawing pink boxes around the area.  

- Notice in the partial image of the settings page below that the app name and version number have boxes drawn over them. This is because the VB version of the same app has a different name (ending VB, not CS) and so creates the difference.  
- The app name in the title bar is not highlighted as a difference because the area where the app title will be displayed is excluded from the difference comparison.

![image](https://user-images.githubusercontent.com/189547/35037867-b3e10592-fb70-11e7-90d6-b23af719e212.png)

Each test takes about 75 seconds to run, so not fast. (See example timings below.)  
However, it much faster and less error-prone (both in terms of project generation and spotting differences) than doing the same manually.   
_Just comparing all single page apps manually will require generating 132 projects, plus the test projects, and then they need to be compiled and run!_

![image](https://user-images.githubusercontent.com/189547/35037480-932e65c0-fb6f-11e7-93cb-d7e812f9ad5b.png)

These tests will continue to be manual only as they can't run headless and need manual verification as there's lots of opportunity for false negatives (reported fails that aren't actually fails.)  
Ideally, in the future, we will have these tests be executed automatically on a central server and then have manual verification of the results.

**Comment by mrlacey** [comment]
These tests also have the added benefit of confirming the generated app can run without error. Just because it compiles doesn't mean it will run (e.g XAML parse issues.) WACK tests would catch anything at an app level but doesn't access every page in the generated app.

**Comment by mrlacey** [comment]
See these test in action:

[![Watch on YouTube](https://img.youtube.com/vi/zu56kQAhLUQ/2.jpg)](https://www.youtube.com/watch?v=zu56kQAhLUQ)

**Comment by ralarcon** [comment]
Looks really great!

@mrlace When consider, we can try to execute automatically in a build machine (like the WACK -will need an interactive agent) and generate a report to be reviewed manually... this can be done on daily basics...

**Comment by mrlacey** [comment]
This work has been more complicated than originally expected and the challenges of comparing screenshots have highlighted some previously unconsidered difficulties.  

There are now some automated UI tests for five pages. Just having these few tested has already identified some issues and will continue to be beneficial in the future. 

A new issue (#1717) has been raised to cover the work for adding similar tests for the other pages once the blocking issues have been addressed or appropriate solutions to the particular challenge of testing those pages are identified.

Will close this issue once the appropriate PR (coming soon) is merged with a view to raising more issues, and work to add further tests, in the future. The current scope of this issue is very broad and could otherwise be around indefinitely while ever more tests are created.

**Comment by ralarcon** [comment]
This work will continue during next release. Moving it.

**Comment by mrlacey** [comment]
#1717 already raised for follow-on work
